### PR TITLE
SSR: Add initial support for placeholder constraints

### PR DESCRIPTION
**Pull request by davidlattimore**

### CODE REVIEW

**Comment by davidlattimore** [comment]
I was originally going to include the type constraint with this PR, but I'm not happy with that constraint yet, so separated this out.

File: crates/ra_ssr/src/matching.rs [hunk 1]
```diff
@@ -515,6 +537,21 @@ impl SsrPattern {
     }
 }
 
+impl NodeKind {
+    fn matches(&self, node: &SyntaxNode) -> Result<(), MatchFailed> {
+        let ok = match self {
+            Self::Literal => {
+                mark::hit!(literal_constraint);
```
Comment (matklad) [inline crates/ra_ssr/src/matching.rs:544]: :+1:

File: crates/ra_ssr/src/parsing.rs [hunk 2]
```diff
@@ -205,23 +220,90 @@ fn tokenize(source: &str) -> Result<Vec<Token>, SsrError> {
...
+                    tokens.next().ok_or_else(|| SsrError::new("Unexpected end of placeholder"))?;
+                if token.kind == SyntaxKind::IDENT {
+                    name = Some(token.text);
+                }
+                loop {
+                    let token = tokens
+                        .next()
+                        .ok_or_else(|| SsrError::new("Placeholder is missing closing brace '}'"))?;
+                    match token.kind {
+                        SyntaxKind::COLON => {
```
Comment (matklad) [inline crates/ra_ssr/src/parsing.rs:240]: You can use `T![:]` here an for other tokens (`T!['{']` for curly, to not break macro syntax)

**Pull request merged**